This is a project in which I will apply my data cleaning skills.
The dataset is used here is from eBay Kleinanzeigen, a classified section of the German eBay website. 

This is modified dataset from Dataquest for 2 reasons here: 
* They sampled 50,000 data points from the full dataset, to ensure your code runs quickly in their hosted environment
* They dirtied the dataset a bit to more closely resemble what I would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

And here is a link a orginal dataset:  here

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv',encoding = 'Latin-1')

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

I will use the columns method to print an array of the existing column names

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Then copy to create a new dictionary with new column names:
* yearOfRegistration to registration_year
* monthOfRegistration to registration_month
* notRepairedDamage to unrepaired_damage
* dateCreated to ad_created
* The rest of the column names from camelcase to snakecase


In [6]:
new_lable = {'dateCrawled':'date_crawled', 'name':'name', 'seller':'seller', 'offerType':'offer_type', 'price':'price', 'abtest':'abtest',
       'vehicleType':'vehicle_type', 'yearOfRegistration':'registration_year', 'gearbox':'gearbox', 'powerPS':'powerPS', 'model':'model',
       'odometer':'odometer', 'monthOfRegistration':'registration_month', 'fuelType':'fuel_type', 'brand':'brand',
       'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created', 'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code',
       'lastSeen':'last_seen'}

In [7]:
autos.rename(new_lable, axis = 1, inplace = True)

Here the changes I have made

In [8]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

#### Basic Data Exploration for more information needed for further cleaning

I will take a look at the descriptive statistics for all columns

In [10]:
autos.describe(include = 'all') 

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

First we look for the columns needed to be done some folowings: 
* Drop any columns that have the repeated values because they are useless for further analysis 
* Columns stored as text that need to be converted into numeric type

Now take a closer look for each individual columns 

In [11]:
autos.loc[:10,:]

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   

    seller offer_type   price   abtest vehicle_type  registration_year  \
0   privat    Angebot  $5,000  control          bus               2004   
1   privat    Angebot  $8,500  control    limousine               1997   
2   privat    Angebot  $8,990     test    limousine               2009   
3   privat    Angebot  $4,350  control   kleinwagen               2007   
4   privat    Angebot  $1,350     test        kombi               2003   
5   privat    Angebot  $7,900     test          bus               2006   
6   privat    Angebot    $300     test    limousine               1995   
7   privat    Angebot  $1,990  control    limousine               1998   
8   privat    Angebot    $250     test          NaN               2000   
9   privat    Angebot    $590  control          bus               1997   
10  privat    Angebot    $999     test          NaN               2017   

      gearbox  powerPS    model   odometer  registration_month fuel_type  \
0     manuell      158   andere  150,000km                   3       lpg   
1   automatik      286      7er  150,000km                   6    benzin   
2     manuell      102     golf   70,000km                   7    benzin   
3   automatik       71   fortwo   70,000km                   6    benzin   
4     manuell        0    focus  150,000km                   7    benzin   
5   automatik      150  voyager  150,000km                   4    diesel   
6     manuell       90     golf  150,000km                   8    benzin   
7     manuell       90     golf  150,000km                  12    diesel   
8     manuell        0    arosa  150,000km                  10       NaN   
9     manuell       90   megane  150,000km                   7    benzin   
10    manuell       90      NaN  150,000km                   4    benzin   

         brand unrepaired_damage           ad_created  nr_of_pictures  \
0      peugeot              nein  2016-03-26 00:00:00               0   
1          bmw              nein  2016-04-04 00:00:00               0   
2   volkswagen              nein  2016-03-26 00:00:00               0   
3        smart              nein  2016-03-12 00:00:00               0   
4         ford              nein  2016-04-01 00:00:00               0   
5     chrysler               NaN  2016-03-21 00:00:00               0   
6   volkswagen               NaN  2016-03-20 00:00:00               0   
7   volkswagen              nein  2016-03-16 00:00:00               0   
8         seat              nein  2016-03-22 00:00:00               0   
9      renault              nein  2016-03-16 00:00:00               0   
10  volkswagen              nein  2016-03-14 00:00:00               0   

    postal_code            last_seen  
0         79588  2016-04-06 06:45:54  
1         71034  2016-04-06 14:45:08  
2         35394  2016-04-06 20:15:37  
3         33729  2016-03-15 03:16:28  
4         39218  2016-04-01 14:38:50  
5         22962  2016-04-06 09:45:21  
6         31535  2016-03-23 02:48:59  
7         53474  2016-04-07 03:17:32  
8          7426  2016-03-26 18:18:10  
9    

I noticed that 'price' and 'odometer' columns having numeric data but stored as text because '$', 'km', ','. So next I will convert them to numeric values 

* First I will eliminate '$', 'km' and ',' 

In [12]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','')
autos[['price','odometer']].head()

price odometer
0  5000   150000
1  8500   150000
2  8990    70000
3  4350    70000
4  1350   150000

* Because I have removed '$','km' and ','. Next step, I will convert them into int type

In [13]:
autos[['price','odometer']]=autos[['price','odometer']].astype(int)

* Last step I need to rename columns for not missing any information about monetary unit and measurment unit 

In [14]:
autos.rename({'price' : 'price_$','odometer':'odometer_km'}, axis = 1, inplace = True)

I need more investigation about these columns: 'seller' , 'offer_type', 'unrepaired_damage', 'nr_of_pictures' to check if they have the repeated value or not 

In [15]:
autos['seller'].value_counts(dropna = False)

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [16]:
autos['offer_type'].value_counts(dropna = False)

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [17]:
autos['unrepaired_damage'].value_counts(dropna = False)

nein    35232
NaN      9829
ja       4939
Name: unrepaired_damage, dtype: int64

In [18]:
autos['nr_of_pictures'].value_counts(dropna = False)

0    50000
Name: nr_of_pictures, dtype: int64

There are 3 columns are useless: 'seller' , 'offer_type' and 'nr_of_pictures' because they have mostly one value. Next I will remove them 

In [19]:
#remove columns
autos = autos.drop(['seller','offer_type','nr_of_pictures'],axis = 1)

#### Explore odometer_km and price_$ columns

In [20]:
autos['price_$'].describe().apply(lambda x: format(x,'f'))

count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price_$, dtype: object

In [21]:
autos['price_$'].value_counts().sort_index(ascending = False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price_$, dtype: int64

In [22]:
autos['price_$'].value_counts().sort_index(ascending = True).head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price_$, dtype: int64

At the very top of the price range, there includes prices well over a million, while only a couple dollars at the bottom. One explanations to these abnormally low prices can be attributed to auctions, which may start at 1 dollar. Another explanation is that the price is a place holder for negotiations .

To prevent these values from skewing the data, I will be removing the entries with price less than 1000 and greater than 350000

In [23]:
autos = autos[autos['price_$'].between(1000,350000)]

In [24]:
autos['price_$'].describe()

count     38626.000000
mean       7255.376275
std        9698.439853
min        1000.000000
25%        2200.000000
50%        4350.000000
75%        8950.000000
max      350000.000000
Name: price_$, dtype: float64

After removing the outliers. Our data in price column has changed significantly (the mean dropped from 9840.043760 to 7255.376275, and the standar deviation dropped from 481104.380500 to 9698.439853)

In [25]:
autos['odometer_km'].describe()

count     38626.000000
mean     122778.568840
std       40796.873127
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

There is no problems with 'odometer_km' columns so I won't make any changes 

#### Exploring with date columns

This dataset has 5 date columns:

* date_crawled : added by the crawler
* last_seen : added by the crawler
* ad_created : from the website
* registration_month : from the website
* registration_year : from the website

Columns date_crawled, ad_created, and last_seen have string datatype where as colummns registration_month, registration_year are in integer datatype.

First, I start with string datatypes.

In [26]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

From the first 5 rows I see that the datas have format type  timestamp(Y-m-d h:m:s)

I will convert them into (Y-m-d) type for analyzing conveniently

In [27]:
date_crawled_counts = autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False)

In [28]:
percentage_date_crawled = date_crawled_counts / date_crawled_counts.sum()

In [29]:

pd.DataFrame({'counts': date_crawled_counts , 'per':percentage_date_crawled.map('{:.2%}'.format)})

counts    per
2016-04-03  0.039145  3.91%
2016-03-20  0.038161  3.82%
2016-03-12  0.037384  3.74%
2016-03-21  0.037281  3.73%
2016-04-04  0.036866  3.69%
2016-03-14  0.036633  3.66%
2016-04-02  0.036297  3.63%
2016-03-28  0.035365  3.54%
2016-03-07  0.035132  3.51%
2016-03-19  0.035132  3.51%
2016-04-01  0.034614  3.46%
2016-03-29  0.033967  3.40%
2016-03-15  0.033630  3.36%
2016-03-10  0.033320  3.33%
2016-03-26  0.033112  3.31%
2016-03-30  0.033061  3.31%
2016-03-11  0.032802  3.28%
2016-03-08  0.032621  3.26%
2016-03-22  0.032517  3.25%
2016-03-09  0.032465  3.25%
2016-03-23  0.032206  3.22%
2016-03-27  0.031404  3.14%
2016-03-31  0.031404  3.14%
2016-03-25  0.030523  3.05%
2016-03-17  0.030472  3.05%
2016-03-16  0.029074  2.91%
2016-03-24  0.029022  2.90%
2016-03-05  0.025553  2.56%
2016-03-13  0.016000  1.60%
2016-03-06  0.013877  1.39%
2016-04-05  0.013359  1.34%
2016-03-18  0.012841  1.28%
2016-04-06  0.003262  0.33%
2016-04-07  0.001502  0.15%

In [30]:
pd.DataFrame({'counts': date_crawled_counts , 'per':percentage_date_crawled.map(lambda x: format(x,'.2%'))})

counts    per
2016-04-03  0.039145  3.91%
2016-03-20  0.038161  3.82%
2016-03-12  0.037384  3.74%
2016-03-21  0.037281  3.73%
2016-04-04  0.036866  3.69%
2016-03-14  0.036633  3.66%
2016-04-02  0.036297  3.63%
2016-03-28  0.035365  3.54%
2016-03-07  0.035132  3.51%
2016-03-19  0.035132  3.51%
2016-04-01  0.034614  3.46%
2016-03-29  0.033967  3.40%
2016-03-15  0.033630  3.36%
2016-03-10  0.033320  3.33%
2016-03-26  0.033112  3.31%
2016-03-30  0.033061  3.31%
2016-03-11  0.032802  3.28%
2016-03-08  0.032621  3.26%
2016-03-22  0.032517  3.25%
2016-03-09  0.032465  3.25%
2016-03-23  0.032206  3.22%
2016-03-27  0.031404  3.14%
2016-03-31  0.031404  3.14%
2016-03-25  0.030523  3.05%
2016-03-17  0.030472  3.05%
2016-03-16  0.029074  2.91%
2016-03-24  0.029022  2.90%
2016-03-05  0.025553  2.56%
2016-03-13  0.016000  1.60%
2016-03-06  0.013877  1.39%
2016-04-05  0.013359  1.34%
2016-03-18  0.012841  1.28%
2016-04-06  0.003262  0.33%
2016-04-07  0.001502  0.15%

From the above frequency table, I see that all of the entries were crawled during March 2016 and into April 2016.

In [31]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
                ...   
2016-04-03    0.039455
2016-04-04    0.037281
2016-04-05    0.011987
2016-04-06    0.003366
2016-04-07    0.001320
Name: ad_created, Length: 74, dtype: float64

It seems that all the ads were created on the website between 2015-June to 2016-April

In [32]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True).apply(lambda x: format(x, '.2%'))

2016-03-05     0.11%
2016-03-06     0.36%
2016-03-07     0.46%
2016-03-08     0.62%
2016-03-09     0.89%
2016-03-10     0.98%
2016-03-11     1.17%
2016-03-12     2.22%
2016-03-13     0.84%
2016-03-14     1.20%
2016-03-15     1.50%
2016-03-16     1.55%
2016-03-17     2.64%
2016-03-18     0.74%
2016-03-19     1.46%
2016-03-20     1.98%
2016-03-21     1.97%
2016-03-22     2.08%
2016-03-23     1.79%
2016-03-24     1.85%
2016-03-25     1.78%
2016-03-26     1.61%
2016-03-27     1.41%
2016-03-28     1.94%
2016-03-29     2.08%
2016-03-30     2.35%
2016-03-31     2.27%
2016-04-01     2.32%
2016-04-02     2.49%
2016-04-03     2.44%
2016-04-04     2.34%
2016-04-05    13.11%
2016-04-06    23.47%
2016-04-07    14.00%
Name: last_seen, dtype: object

It seems that the crawler was fetching data from the site between March and April 2016. However, there is a discrepency in the last three days compared with the previous days

In [33]:
autos['registration_month'].describe()


count    38626.000000
mean         5.998783
std          3.591067
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

The min = 0 in the 'registration_moth' can be known as the unknown registration month for the car. So next, I will convert them as value 'Nan' for clearer expression 

In [34]:
autos.loc[autos['registration_month'] == 0, 'registration_month'] = np.nan 

Here below is the new distribution of registration_month column 

In [35]:
autos['registration_month'].describe()

count    36202.000000
mean         6.400447
std          3.344900
min          1.000000
25%          4.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [36]:
autos['registration_year'].describe()

count    38626.000000
mean      2005.679801
std         86.685138
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

There are two problem in the min value and max value (The minimum year is 1000, when cars never existed, and the maximum year is 9999)

Right now I will eliminate the outliers. I choose the year range from 1896 ( the Germany started registering number plates in 1896) to 2016 (as this data was compiled in 2016)

In [37]:
autos = autos[autos['registration_year'].between(1896,2016)]

In [38]:
autos['registration_year'].describe()

count    37206.000000
mean      2003.969010
std          7.019518
min       1927.000000
25%       2001.000000
50%       2005.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

#### Analyzing prices by brand

In [39]:
autos['brand'].value_counts(normalize = True).map('{:.2%}'.format)

volkswagen        21.08%
bmw               12.54%
mercedes_benz     11.16%
audi               9.76%
opel               8.91%
ford               5.87%
renault            3.73%
peugeot            2.79%
fiat               2.11%
skoda              1.91%
seat               1.73%
smart              1.66%
toyota             1.46%
mazda              1.42%
citroen            1.39%
nissan             1.36%
mini               1.09%
hyundai            1.08%
sonstige_autos     1.05%
volvo              0.90%
kia                0.77%
porsche            0.75%
honda              0.73%
mitsubishi         0.69%
chevrolet          0.66%
alfa_romeo         0.62%
suzuki             0.57%
dacia              0.33%
chrysler           0.32%
jeep               0.28%
land_rover         0.26%
jaguar             0.19%
subaru             0.17%
daihatsu           0.17%
saab               0.14%
daewoo             0.09%
trabant            0.09%
rover              0.07%
lancia             0.07%
lada               0.06%


Next I want to explore the mean price of the top 20 brand 

In [40]:
the_20_brand = autos['brand'].value_counts(normalize = True).map('{:.2%}'.format).head(20)

In [41]:
list_20_brand = the_20_brand.index
list_20_brand

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford', 'renault',
       'peugeot', 'fiat', 'skoda', 'seat', 'smart', 'toyota', 'mazda',
       'citroen', 'nissan', 'mini', 'hyundai', 'sonstige_autos', 'volvo'],
      dtype='object')

I will create a empty dictionary for using loop technique 

In [42]:
mean_price_by_brand = {}

In [43]:
for brand in list_20_brand:
    price = autos.loc[autos['brand'] == brand, 'price_$']
    mean_price_by_brand[brand] = price.mean()
mean_price_by_brand

{'volkswagen': 6645.13260232054,
 'bmw': 9119.20218696398,
 'mercedes_benz': 9302.614402697494,
 'audi': 10322.269347287249,
 'opel': 4219.954737477368,
 'ford': 5331.478021978022,
 'renault': 3590.942321557318,
 'peugeot': 3955.169556840077,
 'fiat': 4008.174744897959,
 'skoda': 6836.696755994359,
 'seat': 5638.640746500778,
 'smart': 3780.4692556634304,
 'toyota': 5573.57169117647,
 'mazda': 5309.526415094339,
 'citroen': 4614.970986460348,
 'nissan': 6428.428007889546,
 'mini': 10715.237037037037,
 'hyundai': 6181.8725,
 'sonstige_autos': 14454.552699228792,
 'volvo': 6151.583832335329}

I will order the dictionary by descending mean price 

In [44]:
sorted(mean_price_by_brand.items(), key= lambda x: x[1], reverse = True)

[('sonstige_autos', 14454.552699228792),
 ('mini', 10715.237037037037),
 ('audi', 10322.269347287249),
 ('mercedes_benz', 9302.614402697494),
 ('bmw', 9119.20218696398),
 ('skoda', 6836.696755994359),
 ('volkswagen', 6645.13260232054),
 ('nissan', 6428.428007889546),
 ('hyundai', 6181.8725),
 ('volvo', 6151.583832335329),
 ('seat', 5638.640746500778),
 ('toyota', 5573.57169117647),
 ('ford', 5331.478021978022),
 ('mazda', 5309.526415094339),
 ('citroen', 4614.970986460348),
 ('opel', 4219.954737477368),
 ('fiat', 4008.174744897959),
 ('peugeot', 3955.169556840077),
 ('smart', 3780.4692556634304),
 ('renault', 3590.942321557318)]

According to the above analysis:

'sonstige_autos' is the most expensive car and 'renault' is the least expensive car in our selected top German brands.

In [45]:
mean_mileage_by_brand = {}

In [46]:
for brand in list_20_brand:
    mileage = autos.loc[autos['brand'] == brand, 'odometer_km']
    mean_mileage_by_brand[brand] = mileage.mean()
mean_mileage_by_brand

{'volkswagen': 125765.65089889073,
 'bmw': 132001.50085763293,
 'mercedes_benz': 130062.6204238921,
 'audi': 127491.04929771413,
 'opel': 123952.92697646348,
 'ford': 119619.96336996337,
 'renault': 121423.93655371305,
 'peugeot': 122341.04046242774,
 'fiat': 107901.78571428571,
 'skoda': 110063.46967559944,
 'seat': 116104.19906687403,
 'smart': 97775.08090614887,
 'toyota': 113740.80882352941,
 'mazda': 119981.1320754717,
 'citroen': 114245.64796905222,
 'nissan': 110335.3057199211,
 'mini': 88679.01234567902,
 'hyundai': 101862.5,
 'sonstige_autos': 89730.07712082262,
 'volvo': 137230.53892215568}

In [47]:
pd.DataFrame({'mean_price':pd.Series(mean_price_by_brand),'mean_mileage':pd.Series(mean_mileage_by_brand)})

mean_price   mean_mileage
volkswagen       6645.132602  125765.650899
bmw              9119.202187  132001.500858
mercedes_benz    9302.614403  130062.620424
audi            10322.269347  127491.049298
opel             4219.954737  123952.926976
ford             5331.478022  119619.963370
renault          3590.942322  121423.936554
peugeot          3955.169557  122341.040462
fiat             4008.174745  107901.785714
skoda            6836.696756  110063.469676
seat             5638.640747  116104.199067
smart            3780.469256   97775.080906
toyota           5573.571691  113740.808824
mazda            5309.526415  119981.132075
citroen          4614.970986  114245.647969
nissan           6428.428008  110335.305720
mini            10715.237037   88679.012346
hyundai          6181.872500  101862.500000
sonstige_autos  14454.552699   89730.077121
volvo            6151.583832  137230.538922

There is also a relationship between mean price and mean mileage. Greater price means lesser mileage

#### Exploring relationship between price of repaired and non repaired cars

In [48]:
autos.groupby('unrepaired_damage')['price_$'].mean()

unrepaired_damage
ja      3899.415517
nein    8067.116247
Name: price_$, dtype: float64

ja and nein means yes/no in English. I will convert these values into English for more readable 

In [49]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].replace(['ja','nein'],['yes','no'])

In [51]:
autos.groupby('unrepaired_damage')['price_$'].mean()

unrepaired_damage
no     8067.116247
yes    3899.415517
Name: price_$, dtype: float64

unrepaired damage cars have the price less than repaired cars (less than 4168$)

#### Conclusion 

In this project, I practiced applying a variety of pandas methods to explore and understand a data set on car listings